modules/widgets/top-posts.php

```diff
@@ -570,37 +573,59 @@ function get_by_views( $count, $args ) {
 			return array();
 		}
 
-		return $this->get_posts( $post_view_ids, $count );
+		return $this->get_posts( $post_view_ids, $count, $types );
 	}
 
-	function get_fallback_posts() {
+	/**
+	 * Get some posts if no posts are found in the stats API
+	 *
+	 * @param int   $count The maximum number of posts to be returned.
+	 * @param array $types The post types that should be returned.
+	 * @return array
+	 */
+	public function get_fallback_posts( $count, $types ) {
 		if ( current_user_can( 'edit_theme_options' ) ) {
 			return array();
 		}
 
-		$post_query = new WP_Query;
+		$post_query = new WP_Query();
+
+		if ( ! is_array( $types ) || empty( $types ) ) {
+			$types = array( 'post', 'page' );
+		}
 
 		$posts = $post_query->query(
 			array(
-				'posts_per_page' => 1,
+				'posts_per_page' => $count,
 				'post_status'    => 'publish',
-				'post_type'      => array( 'post', 'page' ),
+				'post_type'      => $types,
 				'no_found_rows'  => true,
+				'fields'         => 'ids',
 			)
 		);
 
 		if ( ! $posts ) {
 			return array();
 		}
 
-		$post = array_pop( $posts );
-
-		return $this->get_posts( $post->ID, 1 );
+		return $this->get_posts( $posts, $count, $types );
 	}
 
-	function get_posts( $post_ids, $count ) {
+	/**
+	 * Get posts from an array of IDs
+	 *
+	 * @param array $post_ids The post IDs.
+	 * @param int   $count The maximum number of posts to return.
+	 * @param array $types The post types that should be returned.
+	 * @return array
+	 */
+	public function get_posts( $post_ids, $count, $types ) {
```

Comment:
Should we set `$types = array( 'post', 'page' )` to make it an optional parameter?

Reply:
we could. I didn't because this variable is treated here:

https://github.com/Automattic/jetpack/blob/218e6fbe5cf31b212487eae692038cf7125c47b0/modules/widgets/top-posts.php#L244

But no harm doing that

Reply:
I'm reconsidering this.

This method is not called directly by the main widget method, it is passed through `get_by_likes`, `get_by_views` and `get_fallback_posts`. So this would require a chain of optional and useless arguments... 

Both `$count` and `$types` are present in the widgets `$instance`, so a nice refactor would be to use it instead. Not sure if we should do it now... Also, these methods could/should be private (or live somewhere else in jetpack codebase)

I'm always afraid with backward compatibility but this is a widget and the code, I beleive, is self contained and shouldn't be called from outside...

What do you think? quick refactor?

Reply:
I'm okay covering that in a later PR. Could you either open an issue (if you don't have bandwidth) or open a new PR for it?